### PR TITLE
Support releaseNamespace argument with HelmChartInflationGenerator

### LINKED ISSUE #4593: HelmChartInflationGenerator ignores releaseNamespace
**Describe the bug**

According to the [documentation](https://kubectl.docs.kubernetes.io/references/kustomize/builtins/#_helmchartinflationgenerator_) of the helm inflator it shall be possible to set the release namespace of the helm template. 
When running `kustomize build --enable-helm .` with the kustomization.ymal below the releaseNamespace is ignored. 

**Files that can reproduce the issue**

Example:

kustomization.yaml

```
helmChartInflationGenerator:
- chartName: nats
  chartVersion: v0.13.1
  chartRepoUrl: https://nats-io.github.io/k8s/helm/charts
  releaseName: nats
  releaseNamespace: XXXXXX
```

**Expected output**

Expected output e.g. for ConfigMap, namesapce set to XXXXXX

```
apiVersion: v1
data:
  nats.conf: |
    # NATS Clients Port
    port: 4222

    # PID file shared with configuration reloader.
    pid_file: "/var/run/nats/nats.pid"

    ###############
    #             #
    # Monitoring  #
    #             #
    ###############
    http: 8222
    server_name:$POD_NAME
    lame_duck_duration: 120s
kind: ConfigMap
metadata:
  labels:
    app.kubernetes.io/instance: nats
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: nats
    app.kubernetes.io/version: 2.7.2
    helm.sh/chart: nats-0.13.1
  name: nats-config
  namespace: XXXXXX
```

**Actual output**

```
apiVersion: v1
data:
  nats.conf: |
    # NATS Clients Port
    port: 4222

    # PID file shared with configuration reloader.
    pid_file: "/var/run/nats/nats.pid"

    ###############
    #             #
    # Monitoring  #
    #             #
    ###############
    http: 8222
    server_name:$POD_NAME
    lame_duck_duration: 120s
kind: ConfigMap
metadata:
  labels:
    app.kubernetes.io/instance: nats
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: nats
    app.kubernetes.io/version: 2.7.2
    helm.sh/chart: nats-0.13.1
  name: nats-config
  namespace: default
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app.kubernetes.io/instance: nats
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: nats
    app.kubernetes.io/version: 2.7.2
    helm.sh/chart: nats-0.13.1
  name: nats
  namespace: default
spec:
  clusterIP: None
  ports:
  - name: client
    port: 4222
  - name: cluster
    port: 6222
  - name: monitor
    port: 8222
  - name: metrics
    port: 7777
  - name: leafnodes
    port: 7422
  - name: gateways
    port: 7522
  selector:
    app.kubernetes.io/instance: nats
    app.kubernetes.io/name: nats
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: nats-box
    chart: nats-0.13.1
  name: nats-box
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nats-box
  template:
    metadata:
      labels:
        app: nats-box
    spec:
      containers:
      - command:
        - tail
        - -f
        - /dev/null
        env:
        - name: NATS_URL
          value: nats
        image: natsio/nats-box:0.8.1
        imagePullPolicy: IfNotPresent
        name: nats-box
        resources: null
        volumeMounts: null
      volumes: null
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  labels:
    app.kubernetes.io/instance: nats
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: nats
    app.kubernetes.io/version: 2.7.2
    helm.sh/chart: nats-0.13.1
  name: nats
  namespace: default
spec:
  podManagementPolicy: Parallel
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/instance: nats
      app.kubernetes.io/name: nats
  serviceName: nats
  template:
    metadata:
      annotations:
        prometheus.io/path: /metrics
        prometheus.io/port: "7777"
        prometheus.io/scrape: "true"
      labels:
        app.kubernetes.io/instance: nats
        app.kubernetes.io/name: nats
    spec:
      containers:
      - command:
        - nats-server
        - --config
        - /etc/nats-config/nats.conf
        env:
        - name: POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        - name: SERVER_NAME
          value: $(POD_NAME)
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        - name: CLUSTER_ADVERTISE
          value: $(POD_NAME).nats.$(POD_NAMESPACE).svc.cluster.local
        image: nats:2.7.2-alpine
        imagePullPolicy: IfNotPresent
        lifecycle:
          preStop:
            exec:
              command:
              - /bin/sh
              - -c
              - nats-server -sl=ldm=/var/run/nats/nats.pid
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /
            port: 8222
          initialDelaySeconds: 10
          periodSeconds: 60
          successThreshold: 1
          timeoutSeconds: 5
        name: nats
        ports:
        - containerPort: 4222
          name: client
        - containerPort: 7422
          name: leafnodes
        - containerPort: 7522
          name: gateways
        - containerPort: 6222
          name: cluster
        - containerPort: 8222
          name: monitor
        - containerPort: 7777
          name: metrics
        resources: {}
        startupProbe:
          failureThreshold: 30
          httpGet:
            path: /
            port: 8222
          initialDelaySeconds: 10
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 5
        volumeMounts:
        - mountPath: /etc/nats-config
          name: config-volume
        - mountPath: /var/run/nats
          name: pid
      - command:
        - nats-server-config-reloader
        - -pid
        - /var/run/nats/nats.pid
        - -config
        - /etc/nats-config/nats.conf
        image: natsio/nats-server-config-reloader:0.6.2
        imagePullPolicy: IfNotPresent
        name: reloader
        resources: null
        volumeMounts:
        - mountPath: /etc/nats-config
          name: config-volume
        - mountPath: /var/run/nats
          name: pid
      - args:
        - -connz
        - -routez
        - -subz
        - -varz
        - -prefix=nats
        - -use_internal_server_id
        - http://localhost:8222/
        image: natsio/prometheus-nats-exporter:0.9.1
        imagePullPolicy: IfNotPresent
        name: metrics
        ports:
        - containerPort: 7777
          name: metrics
        resources: {}
      shareProcessNamespace: true
      terminationGracePeriodSeconds: 120
      volumes:
      - configMap:
          name: nats-config
        name: config-volume
      - emptyDir: {}
        name: pid
  volumeClaimTemplates: null
---
apiVersion: v1
kind: Pod
metadata:
  annotations:
    helm.sh/hook: test
  labels:
    app.kubernetes.io/instance: nats
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: nats
    app.kubernetes.io/version: 2.7.2
    helm.sh/chart: nats-0.13.1
  name: nats-test-request-reply
spec:
  containers:
  - command:
    - /bin/sh
    - -ec
    - |
      nats reply -s nats://$NATS_HOST:4222 'name.>' --command "echo 1" &
    - |
      "&&"
    - |
      name=$(nats request -s nats://$NATS_HOST:4222 name.test '' 2>/dev/null)
    - |
      "&&"
    - |
      [ $name = test ]
    env:
    - name: NATS_HOST
      value: nats
    image: synadia/nats-box
    name: nats-box
  restartPolicy: Never
```

**Kustomize version**
`{Version:kustomize/v4.5.3 GitCommit:de6b9784912a5c1df309e6ae9152b962be4eba47 BuildDate:2022-03-24T20:51:20Z GoOs:linux GoArch:amd64}`

**Platform**
Linux

**Additional context**

It seems that the ReleaseNamespace arguments are missing in the mapping between `HelmChartArgs` and `HelmChart`  [see](https://github.com/kubernetes-sigs/kustomize/blob/cf3a452ddd6f83945d39d582243b8592ec627ae3/api/types/helmchartargs.go#L113)

```
func makeHelmChartFromHca(old *HelmChartArgs) (c HelmChart) {
	c.Name = old.ChartName
	c.Version = old.ChartVersion
	c.Repo = old.ChartRepoURL
	c.ValuesFile = old.Values
	c.ValuesInline = old.ValuesLocal
	c.ValuesMerge = old.ValuesMerge
	c.ReleaseName = old.ReleaseName
	return
}
```

### PR DESCRIPTION
**Pull request by pxwxnvermx**
Resolves #4593 

Signed-off-by: Pawan Verma <pawanjbs5@gmail.com>

### CODE REVIEW

**Comment by wibed** [comment]
@KnVerey @yuwenma any update on this?

**Comment by yuwenma** [comment]
IIRC, kustomize is aimed at putting the built-in functions in a curated and stable status, and not expecting adding new features. Thus, it provides a `render-helm-charts` function which is more flexible to change.  I'm not sure if we should extend this builtin helm or make the change in the other function. Maybe Katrina can give more context and guides here.

As for the helm release namespace config itself, by definition this field would set the namespace for a release, which conflicts with the kustomize namespace transformer. I think supporting this feature is sort of forcing kustomize to make another exception for helm which opens more black holes for other users. @pxwxnvermx please correct me if I misunderstand the use cases here. FYI, in which case that you need both kustomize and helm?

**Comment by psigen** [comment]
@yuwenma : I am not sure about the PR author, but [in the original issue](https://github.com/kubernetes-sigs/kustomize/issues/4593), you can see a number of folks who have use cases that are affected.

Generally, people (myself included) are using [HelmChartInflationGenerator](https://kubectl.docs.kubernetes.io/references/kustomize/builtins/#_helmchartinflationgenerator_) as a lightweight mechanism to embed well-supported Helm Charts in otherwise minimal installations.  For example, if I create a test environment where I test a small application against the `latest` bitnami Helm Chart for a complex service e.g. Mongo.  It might only be two or three manifests, and then the inflated helm chart.

This _almost_ works, and is massively easier to create and maintain than implementing an entire CI workflow to render the helm chart in a pre-build step -- except for this one field not being propagated, which creates a non-functional rendering.

Perhaps you can comment on when there are situations where passing this field through would create a conflict?  If it is only potentially problematic in cases where someone is using the `HelmChartInflationGenerator`, then it would only affect use-cases which are already broken, so it would be better than the current state.

**Comment by KnVerey** [comment]
@yuwenma while you're right to be concerned about adding features to the Helm transformer, this change does fit within our long-term support plan, which you can read here: https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/helmcharts/#long-term-support. Please note that we also accepted the linked issue in this case: https://github.com/kubernetes-sigs/kustomize/issues/4593#issuecomment-1183764375. Please do review the linked issue and this PR when you have a chance.

File: api/krusty/helmchartinflationgenerator_test.go [hunk 1]
```diff
@@ -233,3 +233,283 @@ spec:
   type: ClusterIP
 `)
 }
+
+func TestHelmChartInflationGeneratorWithNamespace(t *testing.T) {
```
Comment (yuwenma) [inline api/krusty/helmchartinflationgenerator_test.go:237]: Regarding [my question](https://github.com/kubernetes-sigs/kustomize/pull/4716#issuecomment-1251604604) about the potential conflict with kustomize's own `namespace` transformer. Can you add another unit test(s) to cover the following cases:
###  `namespace` configured by both kustomize and helm 
```yaml
# kustomization.yaml  
namespace: dev1 # managed-by-kustomize
helmChartInflationGenerator:
- chartName: nats
  chartVersion: v0.13.1
  chartRepoUrl: https://nats-io.github.io/k8s/helm/charts
  releaseName: nats
  releaseNamespace: dev2 # managed-by-helm
```
###  Base helm `namespace`, overlay kustomize `namespace`
@KnVerey Shall kustomize support this case? 
```yaml
# base/kustomization.yaml  
helmChartInflationGenerator:
- chartName: nats
  chartVersion: v0.13.1
  chartRepoUrl: https://nats-io.github.io/k8s/helm/charts
  releaseName: nats
  releaseNamespace: base # managed-by-helm
```
```yaml
# dev/kustomization.yaml  
resources: ../base
namespace: dev # managed-by-kustomize
```
###  Base kustomize `namespace`, overlay helm `namespace`
@KnVerey Shall kustomize support this case? 
```yaml
# base/kustomization.yaml  
namespace: base # managed-by-kustomize
```
```yaml
# dev/kustomization.yaml  
resources: ../base
helmChartInflationGenerator:
- chartName: nats
  chartVersion: v0.13.1
  chartRepoUrl: https://nats-io.github.io/k8s/helm/charts
  releaseName: nats
  releaseNamespace: dev # managed-by-helm
```

**Comment by yuwenma** [comment]
@KnVerey Thanks for the links. Yeah, the short term and long term plans are very reasonable. 
@pxwxnvermx Sorry for the confusions. reviewed 😃 !

File: api/krusty/helmchartinflationgenerator_test.go [hunk 2]
```diff
@@ -233,3 +233,283 @@ spec:
...
+- chartName: nats
+  chartVersion: v0.13.1
+  chartRepoUrl: https://nats-io.github.io/k8s/helm/charts
+  releaseName: nats
+  releaseNamespace: test
+`)
+
+	m := th.Run(th.GetRoot(), th.MakeOptionsPluginsEnabled())
+
+	th.AssertActualEqualsExpected(m, `
```
[the remark this answers is not in the document]
Comment (pxwxnvermx) [reply api/krusty/helmchartinflationgenerator_test.go:255]: I have tested the change with the minecraft chart that is specified in the test case above this one, that chart does not change the namespace on helm namespace and the output does not contain any namespace even with the releaseNamespace argument specified  in kustomization.yaml. Other charts generally have a longer output.

**Comment by schlichtanders** [comment]
@KnVerey @yuwenma any update on this? This fix is important for many kustomize users.

**Comment by jaknowakowski** [comment]
Any update on this?

**Comment by pxwxnvermx** [comment]
I have no update on this PR. I am going to close this as I am not able to focus my time on this PR.

**Pull request closed without merging**